ta/pkcs11/src/pkcs11_attributes.c

```diff
@@ -2478,7 +2484,34 @@ static enum pkcs11_rc set_secret_key_data(struct obj_attrs **head, void *data,
 	if (rc != PKCS11_CKR_OK && rc != PKCS11_RV_NOT_FOUND)
 		return PKCS11_CKR_GENERAL_ERROR;
 
-	return add_attribute(head, PKCS11_CKA_VALUE, data, key_length);
+	rc = add_attribute(head, PKCS11_CKA_VALUE, data, key_length);
+	if (rc != PKCS11_CKR_OK)
+		return rc;
+
+	/* Remove the default empty check value attribute if found */
+	rc = remove_empty_attribute(head, PKCS11_CKA_CHECK_VALUE);
```

Comment:
Should not remove the attribute if present.
If present and of size 0, then `PKCS11_CKA_CHECK_VALUE` should be that one.
If present and of non-0 size, we should compute the value and succeed only if the provided value matches the expected value.

Reply:
@etienne-lms could you clarify which "that one" you are referring to you mean the one calculated from the resulting key ?
For the second case, the specs says that if provided the value provided should be taken even if it conflicts with the expected value. 
Could you please precesie ?

Reply:
Sure.
Extracts from the PKCS#11 Base Specification, section 4.10 Secret key objects:
* _The attribute is optional, but if supported, regardless of how the key object is created or derived, the value of the attribute is always supplied._
* _If a value is supplied in the application template (allowed but never necessary) then, if supported, it MUST match what the library calculates it to be or the library returns a CKR_ATTRIBUTE_VALUE_INVALID._
* _The generation of the KCV may be prevented by the application supplying the attribute in the template as a no-value (0 length) entry._
* _The application can query the value at any time like any other attribute using C_GetAttributeValue._
* _C_SetAttributeValue may be used to destroy the attribute, by supplying no-value._

Reply:
Sorry i was wrong in one point: it seems at object creation, the client template cannot prevent generation of CKA_CHECK_VALUE attribute : _".. it MUST match what the library calculates."_. To destroy it, the client must later use C_SetAttributeValue and set a _"no-value"_ (that is a zero-sized value for the attribute.